### ISSUE TITLE
Add non-throwing LocalDate(Time) construction

### ISSUE THREAD

**Issue by fluidsonic** [Kotlin/kotlinx-datetime#68]
There should be `null`-returning variants for creating `LocalDate(Time)` instances. It shouldn't be necessary to catch an exception. This is especially useful when creating instances from user input.

```kotlin
val date = runCatching { LocalDate(userInput.year, userInput.month, userInput.day) }.getOrNull()
```

**Comment by dkhalanskyjb** [comment]
`orNull` variants of `TimeZone.of` or `Something.parse` probably make sense and follow the spirit of the provided link, which describes `orNull` as a modifier that shows that error conditions are expected (for example, when validating user input) and must be handled.

However, why is this needed for `LocalDate`? Could you please describe some situations where a programmer has values for year, month, and day, then puts these values into a `LocalDate` constructor, which rejects them, but this error is not a logic error and can somehow be recovered from?

**Comment by fluidsonic** [comment]
User input directly from form fields.

**Comment by dkhalanskyjb** [comment]
Wouldn't one want the users to know what field did they make a mistake in? `null` does not say much, so custom logic would be needed anyway to show a helpful error.

Better yet, why not give the users a date picker that prevents invalid dates altogether?

**Comment by fluidsonic** [comment]
>Wouldn't one want the users to know what field did they make a mistake in? null does not say much, so custom logic would be needed anyway to show a helpful error.
>
>Better yet, why not give the users a date picker that prevents invalid dates altogether?

Long-term? Probably. It's more work and probably pointless.
Short-term to get a project started in a fast & pragmatic way? No.
Also, if you rely on the browser's own date picker (HTML5 date input) some browsers may not support that. So users enter the date manually. Still needs to be checked.
Also, people understand e.g. "Your date of birth seems incorrect." just fine. There's no that much that can be wrong in a date.

I found more use-cases in my code where I use [`LocalDateOrNull`](https://github.com/fluidsonic/fluid-time/blob/6caff7a16e8c3d0d34feae9ebfc7fe8fbfbf3e24/sources/LocalDate.kt#L15):
- custom date picker implementation (K/JS)
- parse date from CSV import
- parse date from custom identifier that contains a date
- GraphQL date input parsing (needs custom error message, e.g. `LocalDateOrNull(y,m,d) ?: graphError("message")`)

**Comment by dkhalanskyjb** [comment]
What do you think of `LocalDate.orNull(2024, 1, 40)`?

**Comment by JakeWharton** [comment]
The lack of a verb really throws me. I would expect parseOrNull or tryParse or something. The top-level function will always match the class name/pretend to be a constructor, but a companion function could be imported and used bare without the type qualifier.

**Comment by dkhalanskyjb** [comment]
Definitely, the null-returning parsing function is going to be called `parseOrNull` (https://github.com/Kotlin/kotlinx-datetime/pull/508), `orNull` is for a nullable alternative to *constructors*. The problem we're trying to solve is that, for any `LocalDate.fromOrNull`, or `LocalDate.ofOrNull`, or `LocalDate.createOrNull`, there may be an expectation of the corresponding `LocalDate.from`, `LocalDate.of`, and `LocalDate.create`, which we don't want to introduce.

There isn't really a precedent for this, so we are treading new ground.

**Comment by pablichjenkov** [comment]
The problem i found with ...orNull() like functions, is that they hide the reason behind the error or failure that caused the object not being created. 
Ideally builders should return a Result type that encapsulates the actual value as Success and the failure reason as Error. But then not sure if this makes APIs verbose in general.

**Comment by dkhalanskyjb** [comment]
@pablichjenkov, in principle, sure. In this specific case, do you see different error modes of `LocalDate`/`LocalTime`/etc. constructors that you would actually like to distinguish between in your code?

**Comment by pablichjenkov** [comment]
Hi @dkhalanskyjb,
Seeing this now after almost a year. I think I have to turn my Github notifications on. 
I really don't have a specific error that I could point at, perhaps because I am not a domain expert.
But given the fact that modern software practices do not promote exception throwing, the replacement to exception throwing is using `Result<T, E>` oriented APIs. 
Below code will return `null` and I wouldn't know what the reason is.
```
LocalDate.orNull(2024, 1, 1000)
```
I would prefer, even tho more verbose,
```
val dateResult: DateResult = LocalDate.parse(2024, 1, 40)
localDate LocalDate = dateResult.value ?: run { handleError(dateResult.error) } 
```
In this case the error would be DAY_OUT_OF_RANGE.
